tests/unit/s2n_build_test.c

```diff
@@ -0,0 +1,228 @@
+/*
+* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+#define _GNU_SOURCE
+#include <ctype.h>
+#include <openssl/crypto.h>
+#include <openssl/opensslv.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "s2n_test.h"
+
+struct libcrypto_info {
+    char *name;
+    int major_version;
+    int minor_version;
+    int patch_version;
+    bool is_fips;
+};
+
+void init_libcrypto_info(struct libcrypto_info *li)
+{
+    if (li != NULL) {
+        li->name = NULL;
+        li->major_version = -1;
+        li->minor_version = -1;
+        li->patch_version = -1;
+        li->is_fips = false;
+    }
+}
+
+int number_from_version(char *version)
+{
+    if (version == NULL)
+        return -1;
+    if (version[0] == '0')
+        return 0;
+    int i = atoi(version);
+    /* atoi returns 0 on error */
+    if (i == 0) {
+        return -1;
+    } else {
+        return i;
+    }
+}
+
+int convert_version_number_to_string(struct libcrypto_info *li, char *version_string_out)
+{
+    /* Format major minor and patch version in a normalized version string (e.x. "1.1.1", "3.0", "3", "")
+     * If there is no major version, return -1 and leave version_string_out untouched. */
+    if (li->major_version != -1) {
+        sprintf(version_string_out, "%d", li->major_version);
+        version_string_out = &version_string_out[1];
+    } else {
+        return -1;
+    }
+    if (li->minor_version != -1) {
+        sprintf(version_string_out, ".%d", li->minor_version);
+        version_string_out = &version_string_out[2];
+    }
+    if (li->patch_version != -1) {
+        sprintf(version_string_out, ".%d", li->patch_version);
+    }
+    return 0;
+}
+
+int extract_libcrypto_info_from_s2n_libcrypto(char *s2n_libcrypto_copy, struct libcrypto_info *li)
```

Comment:
I'm somewhat worried about these parsing functions. If we decide to change the name of a build preset, will anyone else be able to edit this test to get it working?

Reply:
The idea behind these parsing functions was to be flexible in the event of future libcryptos. What I had before reflected the processing in `codebuild/bin/s2n_set_build_preset.sh`. But that was messy to. I think @dougch might some idea's about how build presets might change over time. --> As for this code maybe there is a cleaner way to write it with sscanf? Unfortunately the build presets (at least those supported in `codebuild/bin/s2n_set_build_preset.sh`) don't follow an entirely consistent pattern.

Reply:
The other thing to consider is that I think we control the build preset names. So if we want, we could make them a consistent pattern. I'm not super sure if that's possible though.